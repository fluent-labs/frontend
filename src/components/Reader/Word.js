import React from "react";
import PropTypes from "prop-types";

import { Button, Card, List } from "semantic-ui-react";

const Word = props => {
  const {
    language,
    token,
    tag,
    lemma,
    definitions: { definitions }
  } = props;

  let header = token;
  let meta;
  if (language === "CHINESE") {
    const {
      definitions: { pinyin },
      hsk
    } = props;

    if (pinyin && pinyin != null) {
      header += ` (${pinyin.join(", ")})`;
    }

    meta = "";
    if (tag && tag != null) {
      meta += tag;
    }
    if (hsk && hsk != null) {
      meta += ` - HSK: ${hsk}`;
    }
  } else {
    if (lemma && lemma != null) {
      header += ` (${lemma})`;
    }
    meta = tag;
  }

  return (
    <Card>
      <Card.Content>
        <Card.Header>{header}</Card.Header>
        <Card.Meta>{meta}</Card.Meta>
        <Card.Description>
          <List bulleted>
            {definitions &&
              definitions != null &&
              definitions[0].subdefinitions.map(definition => {
                return <List.Header key={definition}>{definition}</List.Header>;
              })}
          </List>
        </Card.Description>
      </Card.Content>
      <Card.Content extra>
        <Button basic color="green">
          Add to vocabulary
        </Button>
      </Card.Content>
    </Card>
  );
};

Word.propTypes = {
  language: PropTypes.string.isRequired,
  token: PropTypes.string.isRequired,
  tag: PropTypes.string,
  lemma: PropTypes.string,
  definitions: PropTypes.array,
  hsk: PropTypes.number,
  pinyin: PropTypes.array
};

export default Word;
