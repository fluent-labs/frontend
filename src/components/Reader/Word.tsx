import React from "react";

import { Button, Card, Item, Placeholder } from "semantic-ui-react";
import { Definition } from "./Definition";
import { ApiClient, DefinitionDTO, WordDTO } from "../../client/api/ApiClient";
import { Component } from "react";

type WordProps = {
  language: string;
  word: WordDTO;
};

interface WordState {
  definitions: Array<DefinitionDTO>;
  submissionState: SubmissionState;
}

type WordDisplayInfo = {
  header: string;
  meta: string;
};

export enum SubmissionState {
  LOADING,
  SUCCESS,
  FAILURE,
}

const getWordInfo = (word: WordDTO): WordDisplayInfo => {
  return {
    header: `${word.token} (${word.lemma})`,
    meta: word.tag,
  };
};

const getChineseWordInfo = (word: WordDTO): WordDisplayInfo => {
  const pinyin =
    word.definitions.length > 0
      ? ` (${word.definitions[0].pronunciation})`
      : "";

  return {
    header: `${word.token}${pinyin}`,
    meta: word.tag,
  };
};

export class Word extends Component<WordProps, WordState> {
  state = {
    definitions: [],
    submissionState: SubmissionState.LOADING,
  };
  client = new ApiClient();

  componentDidMount = () => {
    const { language, word } = this.props;
    this.client
      .getDefinition(language, word.token)
      .then((result) =>
        this.setState({
          definitions: result,
          submissionState: SubmissionState.SUCCESS,
        })
      )
      .catch((e) =>
        this.setState({ submissionState: SubmissionState.FAILURE })
      );
  };

  getCards = () => {
    if (this.state.submissionState === SubmissionState.LOADING) {
      return (
        <Card.Group>
          <Card>
            <Card.Content>
              <Card.Header as="a">
                <Placeholder.Line length="short" />
              </Card.Header>
              <Card.Meta>
                Pronunciation: <Placeholder.Line length="very short" />
              </Card.Meta>
              <Card.Description>
                Definitions:
                <Placeholder.Line />
                <Placeholder.Line />
                <Placeholder.Line />
              </Card.Description>
              <Card.Content extra>
                <Button basic color="green">
                  Add to vocabulary
                </Button>
              </Card.Content>
            </Card.Content>
          </Card>
        </Card.Group>
      );
    } else {
      return (
        <Card.Group>
          {this.state.definitions.map((definition: DefinitionDTO) => {
            return <Definition key={definition.id} definition={definition} />;
          })}
        </Card.Group>
      );
    }
  };

  render = () => {
    const { language, word } = this.props;
    const wordInfo: WordDisplayInfo =
      language === "CHINESE" ? getChineseWordInfo(word) : getWordInfo(word);
    return (
      <Item>
        <Item.Content>
          <Item.Header as="a">{wordInfo.header}</Item.Header>
          <Item.Meta>{wordInfo.meta}</Item.Meta>
          <Item.Description>
            <Card.Group>
              {this.state.definitions.map((definition: DefinitionDTO) => {
                return (
                  <Definition key={definition.id} definition={definition} />
                );
              })}
              {/* {loadingCard} */}
            </Card.Group>
          </Item.Description>
        </Item.Content>
      </Item>
    );
  };
}
