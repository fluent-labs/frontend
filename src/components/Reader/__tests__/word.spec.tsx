import React from "react";
import renderer from "react-test-renderer";

import { Word } from "../Word";

describe("Word", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Word
          key="test"
          word={{
            token: "test",
            processedToken: "word",
            tag: "noun",
            lemma: "word",
            definitions: [],
          }}
          definitions={[{ id: "test", subdefinitions: [], tag: "noun" }]}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
