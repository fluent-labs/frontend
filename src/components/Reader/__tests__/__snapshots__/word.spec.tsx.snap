// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word renders correctly 1`] = `
Array [
  <span
    className="ant-typography ant-typography-success"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "WebkitLineClamp": undefined,
      }
    }
  >
    <u>
      test
    </u>
  </span>,
  <span
    className="ant-typography"
    style={
      Object {
        "WebkitLineClamp": undefined,
      }
    }
  >
     
  </span>,
]
`;
